### PR TITLE
Fix Dockerfile (#172, #156)

### PR DESCRIPTION
**Pull request by abitrolly**
This still doesn't work as expected, because `cht.sh.txt` runs in
interactive mode and gives the prompt.

    Where should cheat.sh be installed [/root/.cheat.sh]?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,15 @@
-FROM alpine:latest
+FROM alpine:3.10
```
Comment (SuperSandro2000) [inline Dockerfile:1]:
```suggestion
FROM alpine:3.11
```

File: Dockerfile [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 RUN apk add --update --no-cache python2 py2-pip py2-gevent \
     py2-flask py2-requests py2-pygments py2-redis \
     py2-cffi py2-icu bash vim gawk sed \
     && apk add --no-cache --virtual build-deps python2-dev \
     build-base git \
     && pip install -r requirements.txt \
-    && sh share/scripts/get-sheets.sh \
     && apk del build-deps
+COPY . /app
+RUN apk add --update --no-cache curl git py2-virtualenv \
```
Comment (SuperSandro2000) [inline Dockerfile:12]: You don't need virtualenv in a Dockerfile.

File: Dockerfile [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
+COPY requirements.txt /app/
 RUN apk add --update --no-cache python2 py2-pip py2-gevent \
     py2-flask py2-requests py2-pygments py2-redis \
     py2-cffi py2-icu bash vim gawk sed \
     && apk add --no-cache --virtual build-deps python2-dev \
     build-base git \
     && pip install -r requirements.txt \
-    && sh share/scripts/get-sheets.sh \
     && apk del build-deps
+COPY . /app
```
Comment (SuperSandro2000) [inline Dockerfile:11]: Add a ``.dockerignore`` file with ``.git`` if the .git is not relevant to not bust caches when doing git operations.

File: Dockerfile (hunk 1 above)
Comment (abitrolly) [reply Dockerfile:1]: Doesn't work.
```
STEP 1: FROM alpine:3.11
STEP 2: WORKDIR /app
--> Using cache 4c3f84f4c87b8742a7c6f9f6abf36656231bd2501de6eefef4912f9669cc5496
STEP 3: COPY requirements.txt /app/
--> 6b2f6454876
STEP 4: RUN apk add --update --no-cache python2 py2-pip py2-gevent     py2-flask py2-requests py2-pygments py2-redis     py2-cffi py2-icu bash vim gawk sed     && apk add --no-cache --virtual build-deps python2-dev     build-base git     && pip install -r requirements.txt     && apk del build-deps
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
ERROR: unsatisfiable constraints:
  py2-cffi (missing):
    required by: world[py2-cffi]
  py2-flask (missing):
    required by: world[py2-flask]
  py2-gevent (missing):
    required by: world[py2-gevent]
  py2-icu (missing):
    required by: world[py2-icu]
  py2-pygments (missing):
    required by: world[py2-pygments]
  py2-redis (missing):
    required by: world[py2-redis]
  py2-requests (missing):
    required by: world[py2-requests]
Error: error building at STEP "RUN apk add --update --no-cache python2 py2-pip py2-gevent     py2-flask py2-requests py2-pygments py2-redis     py2-cffi py2-icu bash vim gawk sed     && apk add --no-cache --virtual build-deps python2-dev     build-base git     && pip install -r requirements.txt     && apk del build-deps": error while running runtime: exit status 7
```

File: Dockerfile (hunk 2 above)
Comment (abitrolly) [reply Dockerfile:12]: `cht.sh.txt` doesn't work without it.
```
DEPENDENCY: "virtualenv" is needed to install cheat.sh in the standalone mode
Error: error building at STEP "RUN apk add --update --no-cache curl git     && ./share/cht.sh.txt --standalone-install /app/.cheat.sh/": error while running runtime: exit status 1
```

File: Dockerfile (hunk 3 above)
Comment (abitrolly) [reply Dockerfile:11]: `.dockerignore` is already present.

File: Dockerfile (hunk 1 above)
Comment (SuperSandro2000) [reply Dockerfile:1]: They moved most of those packages to python3.

**Comment by chubin** [comment]
@abitrolly Wow! Thank you for starting fixing it.
It is a very important fix, and that is very important to switch the `Dockerfile` to use the standalone installer!
Very good.

Please use Python 3, not Python 2! Python 2 support will be dropped at some point

Also, if you need tests to verify your `Dockerfile`, they are in `tests/`

**Comment by SuperSandro2000** [comment]
Alpine 3.11 and 3.12 also have all pip packages for python3 pre compiled.

**Comment by abitrolly** [comment]
@SuperSandro2000 can't find https://pypi.org/project/colored/ and https://pypi.org/project/rapidfuzz/ in Alpine.

**Comment by SuperSandro2000** [comment]
You should be able to just do ``pip3 install colored rappidfuzz``

**Comment by abitrolly** [comment]
I managed to get all dependencies for fetching sheets right. The resulting image is 474MB. Haven't tested if the server works yet. I could not find which port it uses by default.

**Comment by chubin** [comment]
@abitrolly Depending on the usecase it maybe possible to remove all `.git` directories in the upstream reps; it would reduce the image size substantially, but to update the cheat sheets one would need to update the image

**Comment by abitrolly** [comment]
With `dive` I could see two improvements here.

* [ ] make shallow git clones
* [ ] remove pip cache

But for now I need to check that webserver works.

**Comment by chubin** [comment]
@abitrolly Ok, I see that you've managed it with the port, but if not it is 8002,
here are the defaults:

https://github.com/chubin/cheat.sh/blob/master/lib/config.py

+ you can override them with `etc/config.yaml`

+ you can run the tests; you can't find an example how they are started in `cht.sh.txt`
Basically, they run during the installation anyway (if you use `cht.sh.txt` for installation, and you do),
but you can use them to test a service also

File: Dockerfile [hunk 4]
```diff
@@ -1,13 +1,19 @@
 FROM alpine:latest
```
Comment (SuperSandro2000) [inline Dockerfile:1]:
```suggestion
FROM alpine:3.12
```

**Comment by abitrolly** [comment]
I got 500 error when running server.
```
$ podman build . -t cheat.sh
...
$ podman run -it --rm -p 8002 cheat.sh
Starting server on 0.0.0.0:8002
127.0.0.1 - - [2020-07-28 17:23:28] "GET / HTTP/1.1" 500 426 0.036528
127.0.0.1 - - [2020-07-28 17:23:28] "GET /favicon.ico HTTP/1.1" 404 356 0.002009
^CKeyboardInterrupt
```
How to turn on debug info to see the trace?

File: Dockerfile (hunk 1 above)
Comment (abitrolly) [reply Dockerfile]: Switched to Python 3 and 3.12

**Comment by abitrolly** [comment]
Doesn't look like the server is Python 3 ready.
```
Starting server on 0.0.0.0:8002
127.0.0.1 - - [2020-07-29 11:14:05] "GET /favicon.ico HTTP/1.1" 404 356 0.002351
...
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "bin/srv.py", line 272, in answer
    log_query(ip_address, found, topic, user_agent)
  File "bin/srv.py", line 101, in log_query
    my_file.write(log_entry.encode('utf-8')+"\n")
TypeError: can't concat str to bytes
```

**Comment by chubin** [comment]
@abitrolly Probably you are right, but the fix must be trivial; could you please do it? It should be only this line or maybe a little bit more, but not much

**Comment by abitrolly** [comment]
Finally no tracebacks or errors. Although the output is strange.

image

Maybe it is because stylesheets are not found.
```
127.0.0.1 - - [2020-07-29 12:01:14] "GET /files/fonts/stylesheet.css HTTP/1.1" 404 356 0.002657
```

**Comment by chubin** [comment]
No it is not because of the stylesheets; for me it looks like a python 3 string encoding issue, that `\` is somehow eaten, and `\e[0m` becomes `x1b`

**Comment by abitrolly** [comment]
@chubin could we merge it for now and then concentrate on troubleshooting Python 3 issues? At least this PR fixes the build, and the server can be started successfully with it.

**Comment by abitrolly** [comment]
image

`curl` works ok.

**Comment by abitrolly** [comment]
The `html` output is returned by this function.

https://github.com/chubin/cheat.sh/blob/38edd233871abc00db0eb8e06217816941ee62b5/lib/cheat_wrapper.py#L22

**Comment by chubin** [comment]
ah ok, now I understand; it was a screenshot of a browser; I saw `curl` there, and so I automatically thought it was terminal, now I see.

Yes, this PR can be merged

**Pull request merged**